provider/gateway/rest/router.go

```diff
@@ -136,9 +142,234 @@ func newRouter(log log.Logger, addr sdk.Address, pclient provider.Client) *mux.R
 		leaseServiceStatusHandler(log, pclient.Cluster())).
 		Methods("GET")
 
+	// POST /lease/<lease-id>/shell
+	lrouter.HandleFunc("/shell",
+		leaseShellHandler(log, pclient.Manifest(), pclient.Cluster()))
+
 	return router
 }
 
+type channelToTerminalSizeQueue <-chan remotecommand.TerminalSize
+
+func (sq channelToTerminalSizeQueue) Next() *remotecommand.TerminalSize {
+	v, ok := <-sq
+	if !ok {
+		return nil
+	}
+
+	return &v // Interface is dumb and use a pointer
+}
+
+type leaseShellResponse struct {
+	ExitCode int    `json:"exit_code"`
+	Message  string `json:"message,omitempty"`
+}
+
+func leaseShellHandler(log log.Logger, mclient pmanifest.Client, cclient cluster.Client) http.HandlerFunc {
+	return func(rw http.ResponseWriter, req *http.Request) {
+		leaseID := requestLeaseID(req)
+
+		//  check if deployment actually exists in the first place before querying kubernetes
+		active, err := mclient.IsActive(context.Background(), leaseID.DeploymentID())
+		if err != nil {
+			log.Error("failed checking deployment activity", "err", err)
+			rw.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		if !active {
+			log.Info("no active deployment", "lease", leaseID)
+			rw.WriteHeader(http.StatusNotFound)
+			return
+		}
+
+		localLog := log.With("lease", leaseID.String(), "action", "shell")
+
+		vars := req.URL.Query()
+		var cmd []string
+
+		for i := 0; true; i++ {
+			v := vars.Get(fmt.Sprintf("cmd%d", i))
+			if 0 == len(v) {
+				break
+			}
+			cmd = append(cmd, v)
+		}
+
+		if len(cmd) == 0 {
+			localLog.Error("missing cmd parameter")
+			rw.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		tty := vars.Get("tty")
+		if 0 == len(tty) {
+			localLog.Error("missing parameter tty")
+			rw.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		isTty := tty == "1"
+
+		service := vars.Get("service")
+		if 0 == len(service) {
+			localLog.Error("missing parameter service")
+			rw.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		stdin := vars.Get("stdin")
+		if 0 == len(stdin) {
+			localLog.Error("missing parameter stdin")
+			rw.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		connectStdin := stdin == "1"
+
+		podIndexStr := vars.Get("podIndex")
+		if len(podIndexStr) == 0 {
+			localLog.Error("missing parameter podIndex")
+			rw.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		var podIndex uint
+		_, err = fmt.Sscanf(podIndexStr, "%d", &podIndex)
+		if err != nil {
```

Comment:
can use [`strconv.ParseUint`](https://golang.org/pkg/strconv/#ParseUint) here.

Also, both of these would be a request error of some sort I think.